lib/smppex/session.ex

```diff
@@ -136,27 +134,25 @@ defmodule SMPPEX.Session do
       {:ok, session, pdus} ->
         new_state = do_send_pdus(%{state | session: session}, pdus)
         parse_pdus(new_state, rest_data)
-      {:stop, session, pdus} ->
+      {:stop, session, pdus, reason} ->
         new_state = do_send_pdus(%{state | session: session}, pdus)
-        do_stop(new_state)
-      :stop ->
-        do_stop(state)
+        do_stop(new_state, reason)
+      {:stop, reason} ->
+        do_stop(state, reason)
     end
   end
 
   defp handle_socket_closed(state) do
-    SMPPHandler.handle_socket_closed(state.session)
-    do_stop(state)
+    do_stop(state, :socket_closed)
   end
 
   defp handle_socket_error(state, reason) do
-    SMPPHandler.handle_socket_error(state.session, reason)
-    do_stop(state)
+    do_stop(state, {:socket_error, reason})
   end
 
-  defp do_stop(state) do
+  defp do_stop(state, reason) do
     _ = state.transport.close(state.socket)
-    SMPPHandler.handle_stop(state.session)
+    SMPPHandler.handle_stop(state.session, reason)
```

Comment:
Have you considered #20? Maybe we could just call `handle_stop` before the `transport.close` above -- that way I check if the reason is not `:socket_closed` and send `unbind` pdust. That way the connection is still usable if the reason isn't socket-related.

Though maybe we should implement `unbind`/`unbind_resp` internally, the same way we automatically handle `enquire_link(_resp)` for the user.

So the rules would be:
- if client sends unbind, send back unbind_resp and immediately shut down
- if we're shutting down, send unbind, and since the code is async, block until unbind_resp or unbind_timeout (3s? by default, our system sends SIGTERM, waits 5s, then SIGKILL, 3s would mean it has a bit of extra time for the rest of shutdown)